lib/logger2.ts

```diff
@@ -0,0 +1,199 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import {Config} from './configParser';
+
+let useChalk = false;
+let chalk: any;
+try {
+  chalk = require('chalk');
+  useChalk = true;
+} catch (e) {
+}
+
+export enum LogLevel {
+  ERROR,
+  WARN,
+  INFO,
+  DEBUG
+}
+
+export enum WriteTo {
+  CONSOLE,
+  FILE,
+  BOTH,
+  NONE
+}
+let logFile = 'protractor.log';
+
+export class Logger {
+  static logLevel: LogLevel = LogLevel.INFO;
+  static showTimestamp: boolean = true;
+  static showId: boolean = true;
+  static writeTo: WriteTo = WriteTo.CONSOLE;
+  static fd: any;
+  static firstWrite: boolean = false;
+
+  static set(config: Config): void {
+    if (config.troubleshoot) {
+      Logger.logLevel = LogLevel.DEBUG;
+    }
+  }
+
+  static setWrite(writeTo: WriteTo, opt_logFile?: string): void {
+    if (opt_logFile) {
+      logFile = opt_logFile;
+    }
+    Logger.writeTo = writeTo;
+    if (Logger.writeTo == WriteTo.FILE || Logger.writeTo == WriteTo.BOTH) {
+      Logger.fd = fs.openSync(path.resolve(logFile), 'a');
+      Logger.firstWrite = false;
+    }
+  }
+
+  constructor(private id: string) {}
+
+  log(logLevel: LogLevel, msgs: any[]): void {
+    switch (Logger.logLevel) {
+      case LogLevel.ERROR:
+        if (logLevel <= LogLevel.ERROR) {
+          this.print_(logLevel, msgs);
+        }
+        break;
+      case LogLevel.WARN:
+        if (logLevel <= LogLevel.WARN) {
+          this.print_(logLevel, msgs);
+        }
+        break;
+      case LogLevel.INFO:
+        if (logLevel <= LogLevel.INFO) {
+          this.print_(logLevel, msgs);
+        }
+        break;
+      case LogLevel.DEBUG:
+        if (logLevel <= LogLevel.DEBUG) {
+          this.print_(logLevel, msgs);
+        }
+        break;
+      default:
+        throw new Error('Log level undefined');
+    }
+  }
+
+  print_(logLevel: LogLevel, msgs: any[]): void {
+    this.printBuffer_(logLevel, msgs);
+  }
+
+  printBuffer_(logLevel: LogLevel, msgs: any[]): void {
+    let consoleLog: string = '';
+    let fileLog: string = '';
+
+    if (Logger.showTimestamp) {
+      consoleLog += Logger.timestamp_(true);
```

Comment:
Hm, I don't like this calling the same thing with true/false booleans. But I can't think of a better way to do it. Formatting is hard :(